lib/wallet/src/Cardano/Wallet/Checkpoints.hs

```diff
@@ -159,195 +149,50 @@ instance Buildable (DeltaCheckpoints a) where
+the function 'extendAndPrune' will prune all checkpoints
+whose block height does not align with the policy.
+It's ok to supply a list of new checkpoints that is denser than required.
 -}
 
--- | Extend the known checkpoints.
-extendCheckpoints
+-- | Extend the known checkpoints and prune unnecessary ones.
+extendAndPrune
     :: (a -> W.Slot)
         -- ^ Convert checkpoint to slot.
     -> (a -> BlockHeight)
         -- ^ Convert checkpoint to block height.
-    -> BlockHeight
-        -- ^ Epoch stability window = length of the deepest rollback.
+    -> CheckpointPolicy
+        -- ^ Policy to use for pruning checkpoints.
     -> BlockHeight
         -- ^ Current tip of the blockchain,
         -- which is *different* from block height of the latest checkpoint.
     -> NE.NonEmpty a
         -- ^ New checkpoints, ordered by increasing @Slot@.
-    -> DeltasCheckpoints a
-extendCheckpoints getSlot getBlockHeight epochStability nodeTip cps =
-    reverse
-        [ PutCheckpoint (getSlot wcp) wcp
-        | wcp <- cpsKeep
-        ]
-  where
-    unstable = Set.map Quantity $ Set.fromList $ sparseCheckpoints cfg nodeTip
-      where
-        -- NOTE
-        -- The edge really is an optimization to avoid rolling back too
-        -- "far" in the past. Yet, we let the edge construct itself
-        -- organically once we reach the tip of the chain and start
-        -- processing blocks one by one.
-        --
-        -- This prevents the wallet from trying to create too many
-        -- checkpoints at once during restoration which causes massive
-        -- performance degradation on large wallets.
-        --
-        -- Rollback may still occur during this short period, but
-        -- rolling back from a few hundred blocks is relatively fast
-        -- anyway.
-        cfg = (defaultSparseCheckpointsConfig epochStability) { edgeSize = 0 }
-    willKeep cp = getBlockHeight cp `Set.member` unstable
-    cpsKeep = filter willKeep (NE.init cps) <> [NE.last cps]
-
--- | Compute a delta to prune the 'Checkpoints'
--- according to 'defaultSparseCheckpointsConfig'.
-pruneCheckpoints
-    :: (a -> BlockHeight)
-        -- ^ Retrieve 'BlockHeight' from checkpoint data.
-    -> BlockHeight
-        -- ^ Epoch stability window = length of the deepest rollback.
-    -> BlockHeight
-        -- ^ Block height of the latest checkpoint.
+        -- See Note [extendAndPrune].
     -> Checkpoints a
+        -- ^ Current checkpoints.
     -> DeltasCheckpoints a
-pruneCheckpoints getHeight epochStability tip (Checkpoints cps) =
-    [ RestrictTo slots ]
+extendAndPrune getSlot getHeight policy nodeTip xs (Checkpoints cps) =
+    prunes ++ additions
   where
-    willKeep cp = getQuantity (getHeight cp) `Set.member` heights
-    slots = Map.keys $ Map.filter willKeep cps
-    heights = Set.fromList $ sparseCheckpoints
-        (defaultSparseCheckpointsConfig epochStability)
-        tip
+    additions = reverse -- largest slot needs to be applied last
```

Comment:
you mean not largest but oldest, correct?

Reply:
Yes. The newest slot also happens to be the numerically largest, but I'll change the wording.